extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriGatewayHandler.java

```diff
@@ -80,16 +81,48 @@ public void handleCommand(ChannelUID channelUID, Command command) {
 
     @Override
     public void initialize() {
+        scheduler.execute(this::initializeAsync);
+    }
+    
+    private synchronized void initializeAsync() {
         TradfriGatewayConfig configuration = getConfigAs(TradfriGatewayConfig.class);
-        if (configuration.host == null || configuration.host.isEmpty()) {
+
+        if (isNullOrEmpty(configuration.host)) {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                     "Host must be specified in the configuration!");
             return;
         }
-        if (configuration.code == null || configuration.code.isEmpty()) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    "Security code must be provided in the configuration!");
-            return;
+        if (isNullOrEmpty(configuration.code)) {
+            if (isNullOrEmpty(configuration.identity) || isNullOrEmpty(configuration.preSharedKey)) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Either security code or identity and pre-shared key must be provided in the configuration!");
+                return;
+            }
+        } else {
+            configuration.identity = UUID.randomUUID().toString().replace("-", "");
+            configuration.preSharedKey = authenticateOnGateway(configuration);
+
+            if (isNullOrEmpty(configuration.preSharedKey)) {
+                if (isOldFirmware()) {
+                    /*
+                     * older firmware - fall back to authentication with security code
+                     * in this case the Thing configuration will not be persisted
+                     */
+                    logger.warn("Gateway with old firmware - please consider upgrading to the latest version.");
+                    configuration.preSharedKey = configuration.code;
+                    configuration.identity = "";
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Pre-shared key was not obtain sucessfully");
+                    return;
+                }
+            } else {
+                logger.info(
+                        "Successfully obtained pre-shared key for identity {}. Storing it in thing configuration and discarding the securiy code.",
+                        configuration.identity);
+
+                configuration.code = null;
+                handleConfigurationUpdate(configuration.getAsMap());
```

Comment:
handleConfigurationUpdate is meant to be called by the framework only.
You especially should not call it if you haven't implemented the method yourself. The default implementation potentially reinitializes your handler, but all you want to do is to quietly set some values.

Reply:
wouldn't ```editConfiguration``` be too expensive in that case ? - The configuration is anyway retrieved earlier from the thing - I don't see a point to get it one more time just to change two values, while I can easily create new ```Configuration``` object from and directly pass it to ```updateConfiguration```

About ```handleConfigurationUpdate``` - it is a leftover from some experiments I did - so OK - will replace it with ```updateConfiguration```

Reply:
Why should `editConfiguration` be more expensive than your `getAsMap()`? Anyhow the [documentation is clear](https://www.eclipse.org/smarthome/documentation/development/bindings/thing-handler.html#updating-the-configuration) on how config updates have to be done.

Reply:
because when using ```editConfiguration``` the sequence is:
1. Create new ```Configuration``` object with the current configuration properties
2. Update the ```Configuration``` object (in the current situation 3 times) to apply all required changes
3. Call ```updateConfiguration``` to store the new configuration properties in the thing's configuration map.

And my proposal is to skip step 2 and have:
1. Create new ```Configuration``` object with the NEW configuration properties
2. Call ```updateConfiguration``` to store the those properties in the thing's configuration map.

OK - I will change this to the way you want it - but I don't think this is the optimal solution

Reply:
Step 1 in your proposal is nothing else than steps 1+2 combined above. You won't get around setting the parameters that you want to change and you better do this on an instance where it is allowed to do it, because it is your own personal instance that you obtained by calling `editConfiguration`and not the shared instance that the ThingManager is maintaining.